In [45]:
# first of all import all the necessary libraries
import pandas as pd
import numpy as np
import plotly.graph_objects as go

In [46]:
# now import both training and testing dataset
training_dataset = pd.read_csv('train.csv')
testing_dataset = pd.read_csv('test.csv')

In [47]:
# display training and check
training_dataset.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [48]:
# display testing data
testing_dataset.head(5)

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [49]:
# display columns of training dataset
training_dataset.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [50]:
# shift survived column to the last
survived = training_dataset.pop('Survived')

In [51]:
# now insert that poped column into the last of the dataset
training_dataset.insert(len(training_dataset.columns), 'Survived', survived)
training_dataset

PassengerId  Pclass                                               Name  \
0              1       3                            Braund, Mr. Owen Harris   
1              2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2              3       3                             Heikkinen, Miss. Laina   
3              4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4              5       3                           Allen, Mr. William Henry   
..           ...     ...                                                ...   
886          887       2                              Montvila, Rev. Juozas   
887          888       1                       Graham, Miss. Margaret Edith   
888          889       3           Johnston, Miss. Catherine Helen "Carrie"   
889          890       1                              Behr, Mr. Karl Howell   
890          891       3                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1    female  38.0      1      0          PC 17599  71.2833   C85        C   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3    female  35.0      1      0            113803  53.1000  C123        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
886    male  27.0      0      0            211536  13.0000   NaN        S   
887  female  19.0      0      0            112053  30.0000   B42        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
889    male  26.0      0      0            111369  30.0000  C148        C   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

     Survived  
0           0  
1           1  
2           1  
3           1  
4           0  
..        ...  
886         0  
887         1  
888         0  
889         1  
890         0  

[891 rows x 12 columns]

In [52]:
# know the shape of your dataset
training_dataset.shape

(891, 12)

so there are 891 rows and 12 colums

In [53]:
# now find null values in the dataset
training_dataset.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
Survived         0
dtype: int64

so there are 177 missing values in age column and 687 missing values in cabin, and 2 missing values in embarked column

In [54]:
# lets explore cabin column and find unique values in it
training_dataset.Cabin.unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [55]:
# lets solve the issue of cabin by splitting it into two columns namely Has_Cabin and No_Cabin
training_dataset['Has_Cabin'] = training_dataset['Cabin'].notna().astype(int)

In [56]:
training_dataset

PassengerId  Pclass                                               Name  \
0              1       3                            Braund, Mr. Owen Harris   
1              2       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2              3       3                             Heikkinen, Miss. Laina   
3              4       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4              5       3                           Allen, Mr. William Henry   
..           ...     ...                                                ...   
886          887       2                              Montvila, Rev. Juozas   
887          888       1                       Graham, Miss. Margaret Edith   
888          889       3           Johnston, Miss. Catherine Helen "Carrie"   
889          890       1                              Behr, Mr. Karl Howell   
890          891       3                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1    female  38.0      1      0          PC 17599  71.2833   C85        C   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3    female  35.0      1      0            113803  53.1000  C123        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
886    male  27.0      0      0            211536  13.0000   NaN        S   
887  female  19.0      0      0            112053  30.0000   B42        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
889    male  26.0      0      0            111369  30.0000  C148        C   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

     Survived  Has_Cabin  
0           0          0  
1           1          1  
2           1          0  
3           1          1  
4           0          0  
..        ...        ...  
886         0          0  
887         1          1  
888         0          0  
889         1          1  
890         0          0  

[891 rows x 13 columns]

In [57]:
# so now we have conversted cabin into binary: 1 means has cabin and 0 means no cabin. Lets now remove cabin column
training_dataset.pop('Cabin')

0       NaN
1       C85
2       NaN
3      C123
4       NaN
       ... 
886     NaN
887     B42
888     NaN
889    C148
890     NaN
Name: Cabin, Length: 891, dtype: object

In [58]:
# as of now we have removed cabin colum. now again see which columns contain missing values
training_dataset.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
Survived         0
Has_Cabin        0
dtype: int64

In [59]:
# find out classes in the dataset
training_dataset['Pclass'].unique()

array([3, 1, 2], dtype=int64)

In [60]:
# now lets find which class has most missing values in age column based on gender
print(training_dataset.groupby(['Pclass', 'Sex'])['Age'].apply(lambda x: x.isna().mean()*100))

Pclass  Sex   
1       female     9.574468
        male      17.213115
2       female     2.631579
        male       8.333333
3       female    29.166667
        male      27.089337
Name: Age, dtype: float64


In [61]:
# lets check data types of all the columns
training_dataset.dtypes

PassengerId      int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
Survived         int64
Has_Cabin        int32
dtype: object

In [62]:
# check the name column
titles = training_dataset.Name.str.extract(' ([A-Za-z]+)\. ', expand = False)
titles.unique()

array(['Mr', 'Mrs', 'Miss', 'Master', 'Don', 'Rev', 'Dr', 'Mme', 'Ms',
       'Major', 'Lady', 'Sir', 'Mlle', 'Col', 'Capt', 'Countess',
       'Jonkheer'], dtype=object)

In [63]:
# i will not drop name column as i wanna use it to create a rank column and then analyze the data based on their ranks
training_dataset['Titles'] = titles

In [64]:
# now drop Name column
training_dataset = training_dataset.drop('Name', axis = 1)

In [65]:
# view dataset
training_dataset.head()

PassengerId  Pclass     Sex   Age  SibSp  Parch            Ticket     Fare  \
0            1       3    male  22.0      1      0         A/5 21171   7.2500   
1            2       1  female  38.0      1      0          PC 17599  71.2833   
2            3       3  female  26.0      0      0  STON/O2. 3101282   7.9250   
3            4       1  female  35.0      1      0            113803  53.1000   
4            5       3    male  35.0      0      0            373450   8.0500   

  Embarked  Survived  Has_Cabin titles  
0        S         0          0     Mr  
1        C         1          1    Mrs  
2        S         1          0   Miss  
3        S         1          1    Mrs  
4        S         0          0     Mr

In [68]:
# now handle sex with one hot encoding
training_dataset = pd.get_dummies(training_dataset, columns = ['Sex'])
training_dataset.head()

PassengerId  Pclass   Age  SibSp  Parch            Ticket     Fare  \
0            1       3  22.0      1      0         A/5 21171   7.2500   
1            2       1  38.0      1      0          PC 17599  71.2833   
2            3       3  26.0      0      0  STON/O2. 3101282   7.9250   
3            4       1  35.0      1      0            113803  53.1000   
4            5       3  35.0      0      0            373450   8.0500   

  Embarked  Survived  Has_Cabin titles  Sex_female  Sex_male  
0        S         0          0     Mr       False      True  
1        C         1          1    Mrs        True     False  
2        S         1          0   Miss        True     False  
3        S         1          1    Mrs        True     False  
4        S         0          0     Mr       False      True

In [70]:
# convert newly created Sex_female and Sex_male into integers
training_dataset.Sex_female = training_dataset['Sex_female'].astype(int)
training_dataset.Sex_male = training_dataset['Sex_male'].astype(int)

In [71]:
# now display training_dataset
training_dataset.head()

PassengerId  Pclass   Age  SibSp  Parch            Ticket     Fare  \
0            1       3  22.0      1      0         A/5 21171   7.2500   
1            2       1  38.0      1      0          PC 17599  71.2833   
2            3       3  26.0      0      0  STON/O2. 3101282   7.9250   
3            4       1  35.0      1      0            113803  53.1000   
4            5       3  35.0      0      0            373450   8.0500   

  Embarked  Survived  Has_Cabin titles  Sex_female  Sex_male  
0        S         0          0     Mr           0         1  
1        C         1          1    Mrs           1         0  
2        S         1          0   Miss           1         0  
3        S         1          1    Mrs           1         0  
4        S         0          0     Mr           0         1

In [73]:
# check dtypes of all columns again
training_dataset.dtypes

PassengerId      int64
Pclass           int64
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
Survived         int64
Has_Cabin        int32
titles          object
Sex_female       int32
Sex_male         int32
dtype: object

In [75]:
# check the unique values in Ticket, Embarked, and titles
training_dataset.Ticket.nunique()

681

In [77]:
# as there are 681 unique values in ticket column so drop it
training_dataset = training_dataset.drop('Ticket', axis = 1)
training_dataset

PassengerId  Pclass   Age  SibSp  Parch     Fare Embarked  Survived  \
0              1       3  22.0      1      0   7.2500        S         0   
1              2       1  38.0      1      0  71.2833        C         1   
2              3       3  26.0      0      0   7.9250        S         1   
3              4       1  35.0      1      0  53.1000        S         1   
4              5       3  35.0      0      0   8.0500        S         0   
..           ...     ...   ...    ...    ...      ...      ...       ...   
886          887       2  27.0      0      0  13.0000        S         0   
887          888       1  19.0      0      0  30.0000        S         1   
888          889       3   NaN      1      2  23.4500        S         0   
889          890       1  26.0      0      0  30.0000        C         1   
890          891       3  32.0      0      0   7.7500        Q         0   

     Has_Cabin titles  Sex_female  Sex_male  
0            0     Mr           0         1  
1            1    Mrs           1         0  
2            0   Miss           1         0  
3            1    Mrs           1         0  
4            0     Mr           0         1  
..         ...    ...         ...       ...  
886          0    Rev           0         1  
887          1   Miss           1         0  
888          0   Miss           1         0  
889          1     Mr           0         1  
890          0     Mr           0         1  

[891 rows x 12 columns]

In [80]:
# find number of unique values in embarked and titles
training_dataset[['Embarked', 'titles']].nunique()

Embarked     3
titles      17
dtype: int64

In [ ]:
# since there are 3 unique values in Embarked column, lets apply one hot encoding
training_dataset = pd.get_dummies(training_dataset, columns = ['Embarked'], )

In [17]:
# find correlation with respect to age
corr_matrix = training_dataset.corr()
corr_matrix

ValueError: could not convert string to float: 'Braund, Mr. Owen Harris'